pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,340 @@
+	allocatable := node.Status.Capacity
+	overheads := calculateDaemonSetOverhead(podlist.Items)
+	if len(node.Status.Allocatable) > 0 {
+		allocatable = node.Status.Allocatable
+		// calculating system daemons overhead
+		for resourceName, quantity := range allocatable {
+			overhead := node.Status.Capacity[resourceName]
+			overhead.Sub(quantity)
+			overheads[resourceName] = overhead
+		}
+	}
+
+	// Populate overhead metrics
+	if err := c.insertGaugeValues(overheads, node, overheadGaugeVec); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate gauge: %w", err)
+	}
+	// Populate total request metrics
+	if err := c.insertGaugeValues(reqs, node, requestsGaugeVec); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate gauge: %w", err)
+	}
+	// Populate total limits metrics
+	if err := c.insertGaugeValues(limits, node, limitsGaugeVec); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate gauge: %w", err)
+	}
+	// Populate allocatable metrics
+	if err := c.insertGaugeValues(allocatable, node, allocatableGaugeVec); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate gauge: %w", err)
+	}
+
+	return nil
+}
+
+func (c *Controller) insertGaugeValues(resources map[v1.ResourceName]resource.Quantity, node *v1.Node, gaugeVec *prometheus.GaugeVec) error {
+
+	for resourceName, quantity := range resources {
+		resourceTypeName := strings.ReplaceAll(strings.ToLower(string(resourceName)), "-", "_")
+		labels, err := c.generateLabels(node, resourceTypeName)
+		// Register the set of labels that are generated for node
+		nodeNamespacedName := types.NamespacedName{Name: node.Name}
+		c.LabelsMap[nodeNamespacedName] = append(c.LabelsMap[nodeNamespacedName], labels)
+		if err != nil {
+			return fmt.Errorf("generate new labels: %w", err)
+		}
+		gauge, err := gaugeVec.GetMetricWith(labels)
+		if err != nil {
+			return fmt.Errorf("generate new gauge: %w", err)
+		}
+		if resourceName == v1.ResourceCPU {
+			gauge.Set(float64(quantity.MilliValue()))
+		} else {
+			gauge.Set(float64(quantity.Value()))
+		}
+	}
+	return nil
+}
+
+func calculateDaemonSetOverhead(pods []v1.Pod) v1.ResourceList {
+	overheads := v1.ResourceList{}
+	// calculating daemonset overhead
+	daemonSetPods := []v1.Pod{}
+	for _, pod := range pods {
+		if pod.GetOwnerReferences()[0].Kind == "DaemonSet" {
+			daemonSetPods = append(daemonSetPods, pod)
+		}
+	}
+	daemonSetRequests, _ := getPodsTotalRequestsAndLimits(daemonSetPods)
+	addResourceQuantity(daemonSetRequests, overheads)
+	return overheads
+}
+
+// GetPodsTotalRequestsAndLimits calculates the total resource requests and limits for the pods.
+// If pod overhead is non-nil, the pod overhead is added to the
+// total container resource requests and to the total container limits which have a non-zero quantity.
+func getPodsTotalRequestsAndLimits(pods []v1.Pod) (reqs map[v1.ResourceName]resource.Quantity, limits map[v1.ResourceName]resource.Quantity) {
```

Comment:
We have helpers for this: 
https://github.com/aws/karpenter/blob/0113a91d7bb2630aaffef2842ac3dc17b4a55517/pkg/utils/resources/resources.go#L30

Reply:
After thinking about this for a while I think we should keep the one that I have here.
First is that the helper function does not return limits. Also I want to calculate the request and limits in one for loop.
Also about the pod.spec.overhead calculation. It is not doing anything now so I think we can keep it. 

Later I think we should update the helper function to also include limits calculation.